src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/Routing/ActionEndpointFactory.cs

```diff
@@ -191,10 +184,14 @@ private static (RoutePattern resolvedRoutePattern, IDictionary<string, string> r
             RequestDelegate requestDelegate = (context) =>
             {
                 var routeData = context.GetRouteData();
-
                 var actionContext = new ActionContext(context, routeData, action);
 
-                var invoker = _invokerFactory.CreateInvoker(actionContext);
+                // We don't want to close over the invokerFactory in ActionEndpointFactory as
+                // that creates cycles in DI. Since we're creating this delegate at startup time
+                // we don't want to create all of the things we use at runtime until the action
+                // actually matches.
+                var invokerFactory = context.RequestServices.GetRequiredService<IActionInvokerFactory>();
```

Comment:
Presumably you could cache this inside the delegate.

Reply:
dids't